storage/client_split_test.go

```diff
@@ -1574,3 +1575,107 @@ func writeRandomTimeSeriesDataToRange(
+// descriptor (and associated BeginTransaction request). Prior to the
+// fix in #9287, holding up the BeginTransaction would allow the
+// updates to meta addressing records to proceed. Because the test
+// performs an initial split at the SystemPrefix, the dist sender ends
+// up issuing a range lookup request after failing to send the entire
+// txn to one range. The range lookup encounters the meta2 intents and
+// trivially aborts the txn, because it doesn't exist. The meta2
+// intents are deleted. We then run a GC which cleans up the aborted
+// txn. When the BeginTransaction proceeds, it succeeds and the rest
+// the the txn runs to completion.
+//
+// This test verifies that the meta records are intact.
+func TestStoreSplitBeginTxnPushMetaIntentRace(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	splitKey := roachpb.RKey("a")
+
+	sCtx := storage.TestStoreContext()
+	sCtx.TestingKnobs.DisableSplitQueue = true
+	store, stopper, manual := createTestStoreWithContext(t, sCtx)
+	defer stopper.Stop()
+
+	// Advance the clock past the transaction cleanup expiration.
+	manual.Increment(storage.GetGCQueueTxnCleanupThreshold().Nanoseconds() + 1)
+
+	// First, create a split after addressing records.
+	args := adminSplitArgs(roachpb.KeyMin, keys.SystemPrefix)
+	if _, pErr := client.SendWrapped(rg1(store), nil, &args); pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	startTime := timeutil.Now()
+	wroteMeta2 := make(chan struct{}, 1)
+	store.TestingKnobs().TestingCommandFilter = func(filterArgs storagebase.FilterArgs) *roachpb.Error {
+		if _, ok := filterArgs.Req.(*roachpb.BeginTransactionRequest); ok {
+			// Return a node unavailable error for BeginTransaction request
+			// in the event we haven't waited 20ms.
+			if timeutil.Now().Sub(startTime) < 20*time.Millisecond {
+				return roachpb.NewError(&roachpb.NodeUnavailableError{})
+			}
+			if log.V(1) {
+				log.Infof(context.TODO(), "allowing BeginTransaction to proceed after 20ms")
+			}
+		} else if filterArgs.Req.Method() == roachpb.Put &&
+			bytes.Equal(filterArgs.Req.Header().Key, keys.RangeMetaKey(splitKey)) {
+			select {
+			case wroteMeta2 <- struct{}{}:
+			default:
+			}
+		}
+		return nil
+	}
+
+	// Initiate split at splitKey in a goroutine.
+	doneSplit := make(chan *roachpb.Error, 1)
+	go func() {
+		if log.V(1) {
+			log.Infof(context.TODO(), "splitting at %s", splitKey)
+		}
+		args := adminSplitArgs(keys.SystemMax, splitKey.AsRawKey())
+		_, pErr := client.SendWrappedWith(store, nil, roachpb.Header{
+			RangeID: store.LookupReplica(splitKey, nil).RangeID,
+		}, &args)
+		doneSplit <- pErr
+	}()
+
+	// Wait for the write to the meta2 key to be initiated.
+	<-wroteMeta2
+
+	// Wait for 5ms to allow meta2 write to finish.
+	time.Sleep(5 * time.Millisecond)
+
+	// GC the replica containing the range descriptor records.
+	if err := store.ManualReplicaGC(store.LookupReplica(splitKey, nil)); err != nil {
+		// TODO(spencer): test fatal here after the gc queue correctly
```

Comment:
If the GC doesn't finish, then the test won't fail as expected before these changes went in.